### PR TITLE
Adds Packaging, build/test/release automation

### PR DESCRIPTION
**Pull request by jvendegna**
Adds the required files to build and release a python package to pypi.
Includes a github action workflow which __should__ trigger a build when you go through the github UI to create a new release.

I took the liberty of moving the tkdesigner and gui dirs into a common parent 'src'. I hope no one minds. If so it can be changed, it's just easier to configure packaging this way IMO.

Added some documentation regarding the effort here and how you can test locally prior to accepting this PR.

Happy to answer any questions about packaging in general, or the CI workflows. I'll point out a few key highlights and todos in the comments. 

Hope it helps! Thanks for inviting me to play!

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        run: |
+          python -m pip install --upgrade pip
+          pip install build
+      - name: Build package
+        run: python -m build
+      - name: Publish package
+        uses: pypa/gh-action-pypi-publish@27b31702a0e7fc50959f5ad993c78deac1bdfc29
+        with:
+          user: __token__
+          password: ${{ secrets.PYPI_API_TOKEN }}
```
[the remark this answers is not in the document]
Comment (jvendegna) [reply .github/workflows/build.yml:52]: obviously I don't need it. Just make sure this env var is in the repo secrets, to be clear.

**Comment by jvendegna** [comment]
Helps with #69

**Comment by jvendegna** [comment]
See: https://test.pypi.org/project/tkdesigner-cli-jvendegna/1.0.0/ for a working package. LMK any thoughts.

File: docs/build.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Building & Releasing to Pypi
```
Comment (jrobchin) [inline docs/build.md:1]: This documentation is great! Much appreciated!

**Comment by jrobchin** [comment]
Personally, instead of having a `src/` directory, I'd prefer having `tkdesigner/` with `gui/` inside it as it is all part of the same tkdesigner package.

**Comment by saur0x** [comment]
I agree on having the directory name `tkdesigner` instead of the `src`. I would also like to propose having a different package called `tkdesigner-gui` for the gui.py module which has the `tkdesigner` package as a dependency.
The cli.py module can be bundled with the `tkdesigner` package as it's just a simple script and doesn't have asset files.

File: .github/workflows/build.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ParthJadhav) [reply .github/workflows/build.yml:52]: Sure, i'll add it once I grasp the whole thing. I am new to GH actions so it might take little time.

**Comment by ParthJadhav** [comment]
Thanks for the amazing additions and changes. 

Yes, I agree with others. We can have two separate packages one for GUI and one for CLI. This way it would be easy to install them. And as of now I am not able to run the GUI through the package and I guess it's not meant to work as of now.

**Comment by neilbaldwin** [comment]
> Thanks for the amazing additions and changes.
> 
> Yes, I agree with others. We can have two separate packages one for GUI and one for CLI. This way it would be easy to install them. And as of now I am not able to run the GUI through the package and I guess it's not meant to work as of now.

Can I suggest something controversial: do we actually need the GUI version (I know, the irony!)? Personally I wouldn't use it now we have a CLI but I'm not a full-time developer so my needs may differ.

**Comment by saur0x** [comment]
> > Thanks for the amazing additions and changes.
> > Yes, I agree with others. We can have two separate packages one for GUI and one for CLI. This way it would be easy to install them. And as of now I am not able to run the GUI through the package and I guess it's not meant to work as of now.
> 
> Can I suggest something controversial: do we actually need the GUI version (I know, the irony!)? Personally I wouldn't use it now we have a CLI but I'm not a full-time developer so my needs may differ.

I think separating out the GUI module from the core package would solve that problem. As one can just install `tkdesigner` package and they will be able to use the `cli.py` module like `python -m tkdesigner.cli <OPTIONS>`.
And novice users can download the `tkdesigner-gui` package and they can use it in a user-friendly way.

**Comment by jvendegna** [comment]
Wow, thanks for the active discussion!

Okay that's a lot so let me break some of this down.

@jrobchin 
> Personally, instead of having a `src/` directory, I'd prefer having `tkdesigner/` with `gui/` inside it as it is all part of the same tkdesigner package.

I could understand that structure, and it may work, but it may require some updates to relative imports. The packages are incredibly small. I am currently staying at a friends way out in the woods in NC, internet is pretty garbage here. I had no trouble uploading and downloading the full package with the GUI included. If that's of concern for anyone.

> Also, I'm assuming the scope of this PR is to setup tests but not aim for full test coverage of the package, correct? If so, I think that's the way to go.

Yeah, best to do test driven development... But no one does, LOL. So we backfill over time. I'm not familiar with the codebase here and am not a professional QA engineer, but I am a professional gatekeeper. And nothing gets packaged and released without some form of tests. I did what I could to get started.

> I don't know too much about this as I usually set it up in IDE, but would it be worth it to add a linting script so that we can run linting locally and possibly add it as a Git hook?

Git hooks are best applied to your own individual local workflow IMO. I would suggest a Makefile instead. This way you can chain commands together under one PHONY target, and they're not hidden in the .git dir. Just an initial thought. Any opinions on this?

@ParthJadhav 
I created #86 so we could track the effort on the PYPI api key. Hope you don't mind.
I am surprised the GUI isn't working. I'll test it too, that's my bad. I thought I did but I must be wrong.

@neilbaldwin 
> Can I suggest something controversial: do we actually need the GUI version (I know, the irony!)? Personally I wouldn't use it now we have a CLI but I'm not a full-time developer so my needs may differ.

I am a full time developer, my first interaction with the cli was perfectly successful and I wouldn't ever use the GUI. But that's just me.

@saur0x 
Thanks for really thinking about this in detail and explaining your position. I can buy into that structure and make it so. I think it makes a lot of sense.

One thing, package name is always the same as the directory the package was built from. So it requires me to rename the gui dir in the project root to tkdesigner-gui. Probably not a deal breaker, but if it is someone tell me asap.

**Comment by jvendegna** [comment]
Well, TIL: You can't publish different packages from the same repository.

So we're back to either:
- tkdesigner package with cli but no gui
- src parent dir for both

**Comment by ParthJadhav** [comment]
> I am currently staying at a friends way out in the woods in NC, internet is pretty garbage here. I had no trouble uploading and downloading the full package with the GUI included. If that's of concern for anyone.

Firstly, Sorry if this is taking up your leisure time.

> @ParthJadhav
> I created #86 so we could track the effort on the PYPI api key. Hope you don't mind.

Sure, No problem.

> I am surprised the GUI isn't working. I'll test it too, that's my bad. I thought I did but I must be wrong.

> > Can I suggest something controversial: do we actually need the GUI version (I know, the irony!)? Personally I wouldn't use it now we have a CLI but I'm not a full-time developer so my needs may differ.

> I am a full time developer, my first interaction with the cli was perfectly successful and I wouldn't ever use the GUI. But that's just me.

>  Well, TIL: You can't publish different packages from the same repository. 
So we're back to either: 
tkdesigner package with cli but no gui 
src parent dir for both


So from above three comments It turns out that GUI can be skipped in the pypi package. We could only have CLI on pypi and GUI could be made available through github.

**Comment by jrobchin** [comment]
What about this:

- cli is packaged with the package and can be interfaced with using the `$ tkdesigner` command
- The GUI is a separate folder outside of the tkdesigner package (separate repo maybe?) that takes tkdesigner as a dependency to generate code

**Comment by jvendegna** [comment]
@ParthJadhav I volunteered my time knowing where I would be when you needed the help. No worries. I enjoy it.

Currently I'm trying to see if there is a way I can structure the project to get everything to just work as a part of a single package. I don't really see a reason you shouldn't be able to either:
`python -m tkdesigner.cli ...`
and
`python -m tkdesigner.gui ...` 
after a pip install.

I think you'd have to alias the command or change the cli.py name to __main__.py to use just 'tkdesigner' to invoke the cli.

If I can't get it to work together I'll pull the GUI out of the package and we can all discuss what more should be done. Publishing it as an independent package will require moving it to another repository.

**Comment by ParthJadhav** [comment]
> @ParthJadhav I volunteered my time knowing where I would be when you needed the help. No worries. I enjoy it.

That's good. Thanks.

> Currently I'm trying to see if there is a way I can structure the project to get everything to just work as a part of a single package. I don't really see a reason you shouldn't be able to either:
> `python -m tkdesigner.cli ...`
> and
> `python -m tkdesigner.gui ...`

I can use the CLI but it's giving me module not found error for gui.

> If I can't get it to work together I'll pull the GUI out of the package and we can all discuss what more should be done. Publishing it as an independent package will require moving it to another repository.

I think Publishing it to a different directory is not ideal.

What do you think ?

**Comment by jvendegna** [comment]
I think there is no way the gui is going to work as a package. I've tried it by now. 

I can include an \_\_init\_\_.py file in the gui/assets directory, but the files will not be included in the package because they are not .py files. Attempting to launch the gui as a module leads to a file not found error regarding the icon and png assets. It just has to be run directly from what I can tell.

So, I think I have a working solution involving just the cli. I'll push up changes and send a package link asap.

**Comment by jvendegna** [comment]
Okay, project is in a working state. There is no way around the fact that the tkdesigner package needs to be contained in a directory. I doubt the name was the problem with src, but I renamed it tkdesigner given that all the source is not in that dir.

So tkdesigner became tkdesigner/tkdesigner. That's python packaging. Don't shoot the messenger. If anyone wants to try to work it out another way, feel free. I feel like I've exhausted what I know to try.

This release is working: https://test.pypi.org/manage/project/tkdesigner-cli-jvendegna/release/1.0.8/

Add notes if you've got them, ask questions. I'll set this aside for a day or two and ensure we're all on the same page as to what changes, if any, we want to see made here.

File: .github/workflows/build.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ParthJadhav) [reply .github/workflows/build.yml:52]: Added the key...

**Comment by ParthJadhav** [comment]
I can't run either GUI or CLI from this commit.
The change in directory may have resulted in this issue. Or maybe I am running it in a wrong way. Can you tell me how are you running it?

**Comment by jvendegna** [comment]
First, thanks for your patience with this. Python packaging is a confusing process, and I may still not have it right.

What I tested was the ability to build, publish, download, and install the tkdesigner module as a package and run the cli module from that package. To recreate that experience:
```sh
mkdir testtkdesigner
cd testtkdesigner
python -m venv .venv
# activate the virtual environment, for windows run activate.bat in the same dir.
.venv/Scripts/activate
pip install -i https://test.pypi.org/simple/ tkdesigner-cli-jvendegna
# For windows just change the variable syntax below or supply the values directly.
python -m tkdesigner.cli $MY_FIGMA_URL $MY_FIGMA_TOKEN -f
```

this does work for me. However, I failed to test the command from the main project and am working on that now. It failed for me too with a module not found error. So, either I need to figure out how to load the right files into the package without the parent tkdesigner directory (this would be my personal preference too, make no mistake, but I'm not sure it's actually possible. I could use a set of helping hands if anyone here has the bandwidth for it.) or we update package imports.

If I update the imports to use `tkdesigner.tkdesigner.cli` and it's dependent references throughout, the cli works locally for me. Again, that is not my preference, but I'm not sure if there is a workaround. 

I'll update the gui references now and see if I can get that working too.

Thanks again, each of you, for your patience, input, and letting me try my hand at this.

**Comment by jvendegna** [comment]
okay, I went to demo why I thought you could not avoid having a parent directory, and I was wrong, and I learned something. Pushing up a lot of reverting changes from that last commit, and I have [published a package](https://test.pypi.org/project/tkdesigner-cli-jvendegna/1.0.9/) without the parent directory and verified the gui is not included and the cli works using the `python -m tkdesigner.cli $URL $TOKEN` command.

Assumptions were made. Reflecting on them I followed the packaging tutorial and assumed the parent was necessary because my configuration was not accurate. In the future I'll ensure that I understand fully prior to communicating those assumptions to others. I hope I've introduced no confusion here. Please let me know if I can clarify anything for anyone.

**Comment by jrobchin** [comment]
We should be able to invoke the cli using just a `tkdesigner` command, no? For example:

`$ tkdesigner FIGMA_URL FIGMA_TOKEN -f`

**Comment by jrobchin** [comment]
Awesome work and learnings by the way!

**Comment by jvendegna** [comment]
> We should be able to invoke the cli using just a `tkdesigner` command, no? For example:
> 
> `$ tkdesigner FIGMA_URL FIGMA_TOKEN -f`

That's a good question. I really don't think so. But I'm not an expert in python packaging. I am enjoying learning, but I'm missing something fundamental if this is true. I know setup.py can be used as an alternative to setup.cfg and can investigate that further.

**Comment by neilbaldwin** [comment]
Probably the best thing is to split this into two separate packages. Tkdesigner which is the CLI version and then tkdesigner-gui which has tkdesigner as one of its dependencies. I’m doing something similar for a small project of mine (which I’m going to rewrite so that I can use tkdesigner!)

**Comment by neilbaldwin** [comment]
If it's any help, I went through the pain of packaging and getting a project onto PyPi with this command-line tool:

https://github.com/neilbaldwin/skb

After going through many, many tutorials I finally settled on this template that I made (I use it for all projects now) as I wanted the least complicated setup that I could get away with that wasn't too clunky.

It uses bumpversion for the versioning and twine for uploading to PyPi.

**Comment by ParthJadhav** [comment]
> okay, I went to demo why I thought you could not avoid having a parent directory, and I was wrong, and I learned something. Pushing up a lot of reverting changes from that last commit, and I have [published a package](https://test.pypi.org/project/tkdesigner-cli-jvendegna/1.0.9/) without the parent directory and verified the gui is not included and the cli works using the `python -m tkdesigner.cli $URL $TOKEN` command.

Nice to hear that. 

> Assumptions were made. Reflecting on them I followed the packaging tutorial and assumed the parent was necessary because my configuration was not accurate. In the future I'll ensure that I understand fully prior to communicating those assumptions to others. I hope I've introduced no confusion here. Please let me know if I can clarify anything for anyone.

Thanks to you we've learned about packaging and it's shortcomings. 

> Everything should work, the GUI and CLI from the source following instructions.md and if you pip install the package linked above you can run the cli, no gui package or module is included.

Awesome !

**Comment by ParthJadhav** [comment]
> I felt fairly certain of what I said when I said it, then I found the [setuptools docs about entry point options](https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html) like 5 mins later 😆.

> I've committed the changes necessary to facilitate invoking tkdesigner directly. I needed to include an entry point to the main cli hook in setup.cfg . I just didn't know that was a thing, to run python modules directly. Makes sense now. pip, twine, etc...

> I published a [working package version 1.0.11 here](https://test.pypi.org/project/tkdesigner-cli-jvendegna/), and it does work for me after pip installing the package into a venv. I'd love to know that it works for someone else too.

Wow ! That's great. It would be far easier to use tkdesigner directly and seems you've accomplished that. Thanks a lot. Going to test it thoroughly. 

Sorry for late response my Internet was half dead since morning.

**Comment by ParthJadhav** [comment]
> If it's any help, I went through the pain of packaging and getting a project onto PyPi with this command-line tool:
> 
> https://github.com/neilbaldwin/skb
> 
> After going through many, many tutorials I finally settled on this template that I made (I use it for all projects now) as I wanted the least complicated setup that I could get away with that wasn't too clunky.
> 
> It uses bumpversion for the versioning and twine for uploading to PyPi.
> 
> Once I had a version of the CLI tool that was installable from PyPi, I then created a GUI wrapper that has the CLI tool as a dependency and made that as a separate package/project. That worked pretty good but I abandoned the GUI version after doing a few more tkinter tutorials so I'm not done with it yet.

Thanks for more information. It would be helpful to @jvendegna

**Comment by jvendegna** [comment]
@neilbaldwin 
> Probably the best thing is to split this into two separate packages. Tkdesigner which is the CLI version and then tkdesigner-gui which has tkdesigner as one of its dependencies. I’m doing something similar for a small project of mine (which I’m going to rewrite so that I can use tkdesigner!)

I tend to agree if you wanted an installable gui package invokable from the cli. However I would also say that is outside of the scope of this effort and should be discussed further with the wider group.

Thanks for the pointers on packaging too, looking into bumpversion now. For reference Kenneth Reitz (author of requests library, pipenv (don't get me started, not everything he touches is pure gold), and more... [has a sweet repo for straight up copy/pasting a setup.py file.](https://github.com/navdeep-G/setup.py). Setuptools dos say static config is preferred, so I'm trying to stick with that.

**Comment by ParthJadhav** [comment]
Yay 🎉 It works !!  Everything including the GUI, CLI and also the PYPI CLI.
https://pypi.org/project/tkdesigner/0.1.1rc0/

Thanks for your Efforts.

If anyone has any pointers they can share now. But I guess everything is good.
Now as everything works, we should shift our focus to CI CD.

**Comment by jvendegna** [comment]
> we should shift our focus to CI CD.
I have most of that lined up in the actions workflow. As you may already know (just don't want to assume your level of knowledge in the area, and may be helpful to someone else) it runs a linter for PEP8 (flake8), a few minimal unit tests, and builds and publishes the package to you.

We could iterate on the jobs. Nail the PEP8 conventions that you want to cover. Write more unit and integration tests. Auto versioning (which I'll work out shortly). I could think of a few others, but repo configuration is on the top of the list if not done already. I'd suggest to ensure that PR's have to pass checks based on the actions to merge to main. [Here are the docs on protecting branches.](https://docs.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests/about-protected-branches)

I'm in no rush to merge, but I will be off to a new location in about a week and am unsure how much more time I'll have to contribute after that. But I should be available to chat and answer some questions.

**Comment by jvendegna** [comment]
I've been reading about [poetry](https://python-poetry.org/) for package management. It has a package.lock file which helps ensure local packages are kept up to date, version, build, and publish commands that are easy to navigate. It basically replaces setup.py, setup.cfg, any versioning mechanism, build, twine, and requirements.txt (which you can still export and supply, just easier to keep accurately updated with poetry it seems).

Does anyone here have experience with it? I think porting the current process over would not take much effort, but I want to know if any of the regular contributors here currently use it. Supplying the requirements.txt as well seems to make it optional.

**Comment by jrobchin** [comment]
I have used poetry in the past and I do think it would be good to move to, but it's not critical. What we have at the moment seems to work and at this point our dependencies are quite simple.

That is to say that go ahead if you'd like, I'm sure it'd be a good learning experience and would make things a little nicer, but definitely not a requirement for this project at this point.

**Comment by jrobchin** [comment]
I think we merge as soon as possible and get the PR validation into place. Adding full test coverage should be broken into smaller issues.

**Comment by ParthJadhav** [comment]
> I have some of that lined up in the actions workflow. As you may already know (just don't want to assume your level of knowledge in the area, and may be helpful to someone else) it runs a linter for PEP8 (flake8), a few minimal unit tests, and builds and publishes the package to you.

> I think we merge as soon as possible and get the PR validation into place. Adding full test coverage should be broken into smaller issues.

Yes, I saw the tests and they are pretty good for now. And I also agree with @jrobchin . We should get the basic tests and build process then we can build over it afterwards.

> We could iterate on the jobs. Nail the PEP8 conventions that you want to cover. Write more unit and integration tests. Auto versioning (which I'll work out shortly). I could think of a few others, but repo configuration is on the top of the list if not done already. I'd suggest to ensure that PR's have to pass checks based on the actions to merge to main. Here are the docs on protecting branches.

I would also suggest to have a separate workflow which are triggered on PR's. These workflow will have to pass basic Syntax Tests and Unit Tests. Off-course we cannot use the  current workflow because it will build and publish the package if the tests are passed on PR's.

> I'm in no rush to merge, but I will be off to a new location in about a week and am unsure how much more time I'll have to contribute after that. But I should be available to chat and answer some questions.

We can merge this PR soon... As you said the To-Do's are :- 

1. Auto Versioning
2. Workflow for PRs
3. Wider Unit Test (Can be added later after this PR is merged)

Let me know If I have missed something...

**Comment by neilbaldwin** [comment]
> I've been reading about [poetry](https://python-poetry.org/) for package management. It has a package.lock file which helps ensure local packages are kept up to date, version, build, and publish commands that are easy to navigate. It basically replaces setup.py, setup.cfg, any versioning mechanism, build, twine, and requirements.txt (which you can still export and supply, just easier to keep accurately updated with poetry it seems).
> 
> Does anyone here have experience with it? I think porting the current process over would not take much effort, but I want to know if any of the regular contributors here currently use it. Supplying the requirements.txt as well seems to make it optional.

I read about Poetry with interest and then jumped over to the Git project page to install it. However a) the install script fails for me and b) nearly 1000 currently outstanding issues on the issue tracker meant I waked away, at least for my current projects. Anyone actually using it? And anyone else have a problem installing it on OSX/zsh (Catalina)?

**Comment by jvendegna** [comment]
@ParthJadhav 

### Branch rules

> Off-course we cannot use the current workflow because it will build and publish the package if the tests are passed on PR's.

I can add a  step rule that the build and publish step only runs on the main branch. Effects: During a PR every commit will trigger the first step of the workflow: linting and testing the codebase. Then when the PR is merged it will re-run the linting and testing job but also run the build and publish steps.

### Auto Versioning

This is more tricky. Poetry makes it pretty simple, but automating versioning should still require some form of manual input, otherwise how to determine if the version bump should be major/minor/patch/rc/etc.... One pattern I've followed professionally to accomplish this is to manually add a git tag to a commit, then use a little one-liner, for example and I'll just use poetry here for simplicity sake `poetry version $(git describe --tags | head -1)` to find the latest tag and bump the version to that. Another option is to include a manually written changelog (I'd argue is extremely valuable) and parse it instead of git tags to find the latest release and bump the package version automagically. While tkdesigner is cool, it isn't mission critical software. We could make a decision to simply bump by patch versions on every PR that hits main, and if you ever wanted to bump the major or minor number... just do it manually. And another option is to just use a commit ref.

@ParthJadhav Let me know how you'd want to maintain this moving forward, and if you have any other ideas I'm open to them for sure.

Let me push up the branch rule change and then I think we could merge. Releasing a new version in the meantime will require a manual update to setup.cfg, and considering the huge changes... is anyone here opposed to v1.0.0? Let me know and I'll push up both changes and this thing will be ready to merge IMO.

**Comment by jvendegna** [comment]
### On Packaging

> I read about Poetry with interest and then jumped over to the Git project page to install it. However a) the install script fails for me and b) nearly 1000 currently outstanding issues on the issue tracker meant I waked away, at least for my current projects. Anyone actually using it? And anyone else have a problem installing it on OSX/zsh (Catalina)?
> 
> Edit: not sure what was going on originally but I have now managed to install Poetry. My comment about 1000 bugs remains though 😆

@neilbaldwin 
I really value your input here. Python packaging has been a contentious issue for me in my career to say the least. My latest role was helping to enable MLOps in a calculus tutoring app. Many team members were of the opinion that Pipenv was the 'officially supported python package manager' and it took me months to get them to understand that it actually wasn't and furthermore our expensive model training machines and their local development practices were burning time waiting on the lock. Wasting a lot of money. At the same time, I didn't know where else too look. Just turned back to pip and requirements.txt and setup.py.

From the research I've done this week, there are downsides to every single option. Static config in setup.cfg can't handle dynamic tasks. Setup.py can seriously screw up your hardware if you invoke it incorrectly. Pipenv's claim to be 'the officially supported python package manager' was just a tagline Kenneth Reitz put in the readme, seems like the dude is a marketing genius. I guess Thea over at Pypa had a discussion with him a while back and that tagline is no longer, after some [heated discussion on reddit](https://www.reddit.com/r/Python/comments/8jd6aq/why_is_pipenv_the_recommended_packaging_tool_by/). 🍿 

All of that being said, remember that GitHub itself has evolved a lot recently too, and issues are no longer 'bugs' in most cases. An issue can be a feature request, documentation nitpick, etc. That being said, there are 540 bug reports in poetry. But respectively pipenv's repo uses too many labels for their issues, which to me obscures the amount of actual bugs. They also have 500 open issues. Setuptools project has more than 500 as well. To me, it's a toss up. [pip itself has 800+ open issues.](https://github.com/pypa/pip/issues) Gotta use something... :shrug:

Given the issue count being relatively similar across the board, I want to choose the one that makes things the easiest. I am deep diving python packaging right now for one reason: to never do it again. It seems the python SIGs have no interest in touching packaging and know its been a mess since day one. They avoid any conversation concerning adding a PEP about it, and I think it's a longshot to expect that to change at this point. So for me, I'm looking for the easiest way I can automate this part of my project setup from here on out. 

Poetry's been around as long as pipenv, has a large community of users, and yes - has its downsides and bugs. But looking through the actual issues a lot of them remind me of #88 here, like - that's not a valid issue, that's a user who refuses to read the docs. And there will never be a shortage of those.

I have been experimenting with poetry since yesterday when I brought it up, and hands down I'll be using it for my projects going forward. And I am happy to migrate this repo to using it if @ParthJadhav wants. If not, what's in place will absolutely suffice. No big deal. 

My reasoning for going all in on poetry.
* Large community
* Active development (last PR to main was a docs update 2 hours ago)
* Uses the actual official standard of configuring your build in pyproject.toml
  * which adds the benefit of eliminating setup.cfg and setup.py and requirements.txt
    * pipenv by contrast adds another required file alongside these - the Pipfile (not suggesting anyone here is advocating pipenv, but I warned you not to get me started :wink:)
    * I want to experiment more with the scripts section of Poetry, but I may be able to remove Makefiles from my templates... like npm scripts.
  * allows developers to ensure they are developing and testing with the same package dependencies
* Manages virtualenvs for you.
* Includes versioning, building, and publishing packages in a single tool.
* Super simple command syntax means I can teach anyone how to package/release.
* includes an init command for new projects, with guided setup automagically creating your pyproject.toml - like npm init and settings.json

I think anyone here who hasn't, should try poetry out and give the workflow a decent test-drive and decide for yourself. Because PYPA is never going to decide for us, and python.org will never commit to solving package management once and for all.

.... and this should be a blog post somewhere... sorry for the wind.

**Comment by neilbaldwin** [comment]
@jvendegna great response Jake.

I've spent the morning making a test project with Poetry and configuring it for building and publishing to PyPi and I have to say it's a breath of fresh air! Ignore my initial scepticism: you're correct that the other popular build systems have as many issues (tracker) and I was aware of that, my comment there was more along the lines of "why would I switch when I'm happy with my current method and Poetry seems just as 'unstable' (poor choice of word)?" but I didn't elucidate that particularly well.

Poetry gets my vote if we're choosing.

**Comment by neilbaldwin** [comment]
> is anyone here opposed to v1.0.0?

No opposition from me.

**Comment by jrobchin** [comment]
Just my 2c, I think Poetry is a good move and the number of issues on a repo speaks far more to its popularity that its stability.

**Comment by jvendegna** [comment]
> @jvendegna great response Jakob.
> 
> I've spent the morning making a test project with Poetry and configuring it for building and publishing to PyPi and I have to say it's a breath of fresh air! Ignore my initial scepticism: you're correct that the other popular build systems have as many issues (tracker) and I was aware of that, my comment there was more along the lines of "why would I switch when I'm happy with my current method and Poetry seems just as 'unstable' (poor choice of word)?" but I didn't elucidate that particularly well.
> 
> Poetry gets my vote if we're choosing.

Thanks mate. Yeah, I'd totally align with anyone who had a system that works well for them not changing it. But like you when I tried poetry I was caught off-guard by the pure simplicity of it. For that reason it has my vote. But this isn't my repo/project, and I don't want to force anything. But since you two are into it I'll get going on a branch we could merge into this and it may be ready by the time Parth chimes in. If he wants it, I'll merge, if not it's good practice migrating an existing repo and I won't mind deleting it. Could introduce a problem or two to work out, but I doubt it would break anything major.

**Comment by ParthJadhav** [comment]
Hello everyone,
It's great to see such active discussion. 

My response might be late... going to read everything, also the Reddit argument.

Learn more about packaging, try, test, build and do all sorts of stuff.

But if it takes long then we should go with poetry. I don't want my lack of experience affect what's best for the project.

Consider this your own repository and do whatever you guys feel the best 😀

**Comment by jvendegna** [comment]
Parth, that's awesome. Let me know if you have any questions or get hung up on anything. I'm in no rush to merge.

**Comment by jvendegna** [comment]
Okay, migration complete. I will push up one more commit that edits some config stuff. Parth, if you want to add my TEST_PYPI_TOKEN as a secret to test publishing the package on the staging server first, I emailed it to you. Then we'll just open another PR to change the build from publishing to testpypi to the real pypi. Cool?

**Comment by ParthJadhav** [comment]
> I can add a step rule that the build and publish step only runs on the main branch. Effects: During a PR every commit will trigger the first step of the workflow: linting and testing the codebase. Then when the PR is merged it will re-run the linting and testing job but also run the build and publish steps.

It works now 🎉

Screenshot 2021-07-24 at 10 21 45 AM

> ### Auto Versioning
> This is more tricky. Poetry makes it pretty simple, but automating versioning should still require some form of manual input, otherwise how to determine if the version bump should be major/minor/patch/rc/etc.... One pattern I've followed professionally to accomplish this is to manually add a git tag to a commit, then use a little one-liner, for example and I'll just use poetry here for simplicity sake `poetry version $(git describe --tags | head -1)` to find the latest tag and bump the version to that. Another option is to include a manually written changelog (I'd argue is extremely valuable) and parse it instead of git tags to find the latest release and bump the package version automagically. While tkdesigner is cool, it isn't mission critical software. We could make a decision to simply bump by patch versions on every PR that hits main, and if you ever wanted to bump the major or minor number... just do it manually. And another option is to just use a commit ref.

Can you give more instructions on that or share me something to read more on it. 

> is anyone here opposed to v1.0.0? Let me know and I'll push up both changes and this thing will be ready to merge IMO.

Yes, we should use version 1.0.0 from here on.

>  Poetry gets my vote if we're choosing.

>  Just my 2c, I think Poetry is a good move and the number of issues on a repo speaks far more to its popularity that its stability.

>  But like you when I tried poetry I was caught off-guard by the pure simplicity of it. For that reason it has my vote.

From all the above votes I guess we should go ahead and use Poetry for Tkinter Designer. I guess we've already done that here https://github.com/ParthJadhav/Tkinter-Designer/pull/84/commits/5760ae47aa678915dafbd0030c0022b1ff6e1fca .

> Parth, that's awesome. Let me know if you have any questions or get hung up on anything. I'm in no rush to merge.

I would like to ask if is there any extra steps to setup the local environment to start working on it ?

Also thanks for the explanation on Poetry vs Pipenv and packaging overall. It was a good read. The reddit discussion was also very interesting. I really liked the line "Refusing that there is a problem is not a way to solve problems." stated here -> https://tinyurl.com/ygp4z43s 

> Okay, migration complete. I will push up one more commit that edits some config stuff. Parth, if you want to add my TEST_PYPI_TOKEN as a secret to test publishing the package on the staging server first, I emailed it to you. Then we'll just open another PR to change the build from publishing to testpypi to the real pypi. Cool? 

Yes, I'll add the token soon...

**Comment by ParthJadhav** [comment]
After Tkinter Designer is live on Real PYPI I'll be posting about Tkinter Designer on Product Hunt 🎉

**Comment by jvendegna** [comment]
> Can you give more instructions on that or share me something to read more on it.

That's a whole bucket. Start by understanding [semantic versioning](https://semver.org/). Then realize there are a trillion ways to peel a potato. Then have a look at [semantic-release](https://semantic-release.gitbook.io/semantic-release/) and how it uses commit message format to automatically determine semantic version bumps. Then have a look at the [other plugins](https://semantic-release.gitbook.io/semantic-release/extending/plugins-list) they offer to handle other aspects of a release... I.e. changelog generation. Then realize that all of this is just one way to skin the potato. IMO auto versioning this way is best for enterprise teams. Not personal pet projects. Too much learning curve for new contributors.

The way I would suggest versioning this project, given all the possibilities, is to manually add a git tag to a commit when you want to create a new release, like so: `git tag -a v1.0.0 -m "tag message"` then `git push --tags`. Read the tags in the pipeline (which is already set up) and  let poetry update the version accordingly. 

Let me know how you want to handle versioning, and I'll make it so.

> Yes, we should use version 1.0.0 from here on.

More precisely we should use 1.0.0 for the merge commit that publishes to pypi. Future updates __should__ update the version number accordingly to semver principals. Breaking changes bump major versions, etc...

> I would like to ask if is there any extra steps to setup the local environment to start working on it ?

Great question. I've included a Makefile here for some quick scripts, and updated it just now to incorporate poetry. That file itself contains documentation on how to use it. But my workflow locally to build and test the package is:

```sh
make setup
# <do code changes>
make setup # installs the package itself into the virtualenv without building, this is just poetry default behavior.
make lint
make test
# set your environment variables for generating a gui if you haven't already
export (set on windows) FIGMA_PROJECT_URL=<whatever>
export FIGMA_TOKEN=<token>
make cli
make gui
# If all that works, commit and push
```

> Also thanks for the explanation on Poetry vs Pipenv and packaging overall. It was a good read.

Anything I can do to demystify and help anyone else understand the trash heap that is python packaging. For now, I'm all in on poetry like "this is the way" and I don't see that changing unless a major security vulnerability is exposed. Thank you for entertaining me  while I fell down the rabbit hole on your project! It really has been great fun. I enjoy working with each of you, and do reach out if any of you want to connect more personally or professionally.

**Comment by neilbaldwin** [comment]
Off topic: thanks to @jvendegna  I've moved the first of my (small) Python projects over to Poetry. So much neater, cleaner and straightforward now. Definitely the (current) future of Python packaging! Thank you Jakob.

**Pull request merged**